File: src/main/java/org/roborace/lapscounter/service/LapsCounterService.java
package org.roborace.lapscounter.service;

import org.roborace.lapscounter.domain.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static java.util.Arrays.asList;
import static java.util.Comparator.comparingInt;
import static java.util.Optional.ofNullable;
import static org.roborace.lapscounter.domain.State.*;

@Service
public class LapsCounterService {

    public static final Logger log = LoggerFactory.getLogger(LapsCounterService.class);

    @Value("${laps.pit-stop-time}")
    private long pitStopTime;

    private State state = READY;
    private long raceTimeLimit = 0;
    private final Stopwatch stopwatch = new Stopwatch();
    private final List<Robot> robots = new ArrayList<>();

    @Autowired
    private FrameProcessor frameProcessor;
    @Lazy
    @Autowired
    private LapsCounterScheduler lapsCounterScheduler;


    public synchronized MessageResult handleMessage(Message message) {

        return switch (message.getType()) {
            case COMMAND -> command(message);
            case STATE -> MessageResult.single(getState());
            case ROBOT_INIT -> robotInit(message);
            case ROBOT_EDIT -> robotEdit(message);
            case ROBOT_REMOVE -> robotRemove(message);
            case TIME -> timeRequest(message);
            case LAPS -> new MessageResult(getLapMessages(robots), ResponseType.SINGLE);
            case LAP_MAN -> lapManual(message);
            case PIT_STOP -> pitStop(message);
            case FRAME -> frame(message);
            default -> throw new LapsCounterException("Method not supported: [" + message.getType() + "]");
        };
    }

    public List<Message> afterConnectionEstablished() {
        return asList(getState(), getTime());
    }

    private MessageResult command(Message message) {
        State parsedState = message.getState();
        if (parsedState == null) {
            throw new LapsCounterException("State is null");
        }
        if (!isCorrectCommand(parsedState)) {
            throw new LapsCounterException("Wrong current state to apply command: [" + state + "]->[" + parsedState + "]");
        }

        MessageResult messageResult = new MessageResult(new ArrayList<>(), ResponseType.BROADCAST);
        if (parsedState != state) {
            state = parsedState;
            messageResult.add(getState());
            switch (state) {
                case READY -> {
                    stopwatch.reset();
                    robots.forEach(Robot::reset);
                    frameProcessor.reset();
                    sortRobotsByLapsAndTime();
                    messageResult.addAll(getLapMessages(robots));
                }
                case STEADY -> {

                }
                case RUNNING -> {
                    stopwatch.start();
                    if (raceTimeLimit > 0) {
                        lapsCounterScheduler.addSchedulerForFinishRace(TimeUnit.SECONDS.toMillis(raceTimeLimit));
                    }
                }
                case FINISH -> {
                    stopwatch.finish();
                    lapsCounterScheduler.resetSchedulers();
                }
            }
        }
        messageResult.add(getTime());
        return messageResult;
    }

    private boolean isCorrectCommand(State newState) {
        if (newState.ordinal() == state.ordinal() + 1) return true;
        return newState.ordinal() == 0 && state.ordinal() == values().length - 1;
    }

    private MessageResult robotInit(Message message) {
        Optional<Robot> existing = getRobot(message.getSerial());
        Robot robot;
        if (existing.isPresent()) {
            log.info("Reconnect robot {}", existing.get());
            robot = existing.get();
        } else {
            robot = new Robot();
            robot.setSerial(message.getSerial());
            robot.setName(ofNullable(message.getName()).orElse("Robot " + message.getSerial()));
            robot.setNum(getNextNum());
            robot.setPlace(robots.size() + 1);
            robot.reset();
            robots.add(robot);
            frameProcessor.robotInit(robot);
            log.info("Connect new robot {}", robot);
        }
        log.debug("Connected robots: {}", robots);
        return MessageResult.broadcast(getLap(robot));
    }

    private int getNextNum() {
        return robots.stream().map(Robot::getNum).max(Integer::compareTo).orElse(0) + 1;
    }

    private MessageResult robotEdit(Message message) {
        Robot robot = getRobotOrElseThrow(message.getSerial());
        log.info("Edit robot {}", robot);
        robot.setName(message.getName());
        return MessageResult.broadcast(getLap(robot));
    }

    private MessageResult robotRemove(Message message) {
        Robot robot = getRobotOrElseThrow(message.getSerial());
        robots.remove(robot);
        frameProcessor.robotRemove(robot);
        MessageResult broadcast = MessageResult.broadcast();
        broadcast.add(message);
        broadcast.addAll(getLapMessages(sortRobotsByLapsAndTime()));
        return broadcast;
    }

    private MessageResult timeRequest(Message message) {
        if (message.getRaceTimeLimit() != null && state != RUNNING) {
            raceTimeLimit = message.getRaceTimeLimit();
            return MessageResult.broadcast(getTime());
        }
        return MessageResult.single(getTime());
    }

    private MessageResult lapManual(Message message) {
        if (state != RUNNING) {
            log.info("Lap manual ignored: state is not running");
            throw new LapsCounterException("Lap manual ignored: state is not running");
        }
        if (message.getLaps() == null) {
            throw new LapsCounterException("Laps count is not setup");
        }

        Robot robot = getRobotOrElseThrow(message.getSerial());
        List<Robot> affectedRobots;
        if (message.getLaps() > 0) {
            affectedRobots = incLaps(robot, stopwatch.getTime());
        } else {
            affectedRobots = decLaps(robot);
        }
        return MessageResult.broadcast(getLapMessages(affectedRobots));
    }

    List<Robot> sortRobotsByLapsAndTime() {
        robots.sort(comparingInt(Robot::getLaps).reversed()
                .thenComparingLong(Robot::getTime)
                .thenComparingInt(Robot::getNum));
        List<Robot> affectedRobots = new ArrayList<>(robots.size());
        for (int i = 0; i < robots.size(); i++) {
            int place = i + 1;
            Robot robot = robots.get(i);
            if (robot.getPlace() != place) {
                robot.setPlace(place);
                affectedRobots.add(robot);
            }
        }
        return affectedRobots;
    }

    private MessageResult pitStop(Message message) {
        if (state != RUNNING) {
            log.debug("PitStop ignored: state is not running");
            throw new LapsCounterException("PitStop ignored: state is not running");
        }
        Robot robot = getRobotOrElseThrow(message.getSerial());

        robot.setPitStopFinishTime(stopwatch.getTime() + pitStopTime);
        log.info("Used PIT_STOP for robot: {}", robot.getSerial());

        Message pitStopStart = Message.builder()
                .type(Type.PIT_STOP)
                .serial(robot.getSerial())
                .build();

        Message pitStopFinish = Message.builder()
                .type(Type.PIT_STOP_FINISH)
                .serial(robot.getSerial())
                .build();
        lapsCounterScheduler.addSchedulerForPitStop(pitStopFinish, pitStopTime);

        return MessageResult.broadcast(asList(pitStopStart, getLap(robot)));
    }

    private MessageResult frame(Message message) {
        if (state != RUNNING) {
            log.debug("Frame ignored: state is not running");
            return null;
        }
        Robot robot = getRobotOrElseThrow(message.getSerial());

        long raceTime = stopwatch.getTime();
        Type frameType = frameProcessor.checkFrame(robot, message.getFrame(), raceTime);
        if (frameType == Type.LAP) {
            List<Robot> robots = incLaps(robot, raceTime);
            return MessageResult.broadcast(getLapMessages(robots));
        } else if (frameType == Type.LAP_MINUS) {
            List<Robot> robots = decLaps(robot);
            return MessageResult.broadcast(getLapMessages(robots));
        } else if (frameType == Type.FRAME) {
            if (frameProcessor.isStartFrame(message.getFrame())) {
                robot.setCurrentLapStartTime(raceTime);
            }
            return MessageResult.single(new Message(Type.FRAME));
        }
        return null;
    }

    private List<Robot> incLaps(Robot robot, long raceTime) {
        robot.incLaps(raceTime);
        List<Robot> affectedRobots = sortRobotsByLapsAndTime();
        if (affectedRobots.isEmpty()) {
            affectedRobots.add(robot);
        }
        return affectedRobots;
    }

    private List<Robot> decLaps(Robot robot) {
        robot.decLaps();
        robot.setTime(robot.extractLastLapTime());
        List<Robot> affectedRobots = sortRobotsByLapsAndTime();
        if (affectedRobots.isEmpty()) {
            affectedRobots.add(robot);
        }
        return affectedRobots;
    }

    private List<Message> getLapMessages(List<Robot> robots) {
        return robots.stream()
                .map(this::getLap)
                .collect(Collectors.toList());
    }

    public Message scheduled() {
        if (state == RUNNING) {
            log.debug("Send time");
            return getTime();
        }
        return null;
    }

    public Message getState() {
        Message message = new Message();
        message.setType(Type.STATE);
        message.setState(state);
        return message;
    }

    public Message getTime() {
        Message message = new Message();
        message.setType(Type.TIME);
        message.setTime(stopwatch.getTime());
        message.setRaceTimeLimit(raceTimeLimit);
        return message;
    }

    private Message getLap(Robot robot) {
        Message message = new Message();
        message.setType(Type.LAP);
        message.setSerial(robot.getSerial());
        message.setName(robot.getName());
        message.setNum(robot.getNum());
        message.setLaps(robot.getLaps());
        message.setTime(robot.getTime());
        message.setBestLapTime(robot.getBestLapTime());
        message.setPitStopFinishTime(robot.getPitStopFinishTime());
        message.setLastLapTime(robot.getLastLapTime());
        message.setPlace(robot.getPlace());
        return message;
    }

    private Robot getRobotOrElseThrow(Integer serial) {
        return ofNullable(serial)
                .flatMap(this::getRobot)
                .orElseThrow(() -> new LapsCounterException("Cannot find robot by serial"));
    }

    private Optional<Robot> getRobot(Integer serial) {
        return robots.stream().filter(r -> r.getSerial() == serial).findAny();
    }
}

File: src/main/kotlin/org/roborace/lapscounter/service/LapsCounterService.kt
package org.roborace.lapscounter.service

import org.roborace.lapscounter.domain.LapsCounterException
import org.roborace.lapscounter.domain.Robot
import org.roborace.lapscounter.domain.State
import org.roborace.lapscounter.domain.Type
import org.roborace.lapscounter.domain.api.Message
import org.roborace.lapscounter.domain.api.MessageResult
import org.roborace.lapscounter.domain.api.MessageResult.Companion.broadcast
import org.roborace.lapscounter.domain.api.MessageResult.Companion.single
import org.roborace.lapscounter.domain.api.ResponseType
import org.roborace.lapscounter.service.util.Stopwatch
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Service
import java.util.concurrent.TimeUnit

@Service
class LapsCounterService(
    private val frameProcessor: FrameProcessor,
    @Value("\${laps.pit-stop-time}") private val pitStopTime: Long = 0
) {

    @Lazy
    @Autowired
    private lateinit var lapsCounterScheduler: LapsCounterScheduler

    private var state = State.READY
    private var raceTimeLimit: Long = 0
    val stopwatch = Stopwatch()
    private val robots: MutableList<Robot> = mutableListOf()


    @Synchronized
    fun handleMessage(message: Message): MessageResult =
        when (message.type) {
            Type.COMMAND -> command(message)
            Type.STATE -> single(getState())
            Type.ROBOT_INIT -> robotInit(message)
            Type.ROBOT_EDIT -> robotEdit(message)
            Type.ROBOT_REMOVE -> robotRemove(message)
            Type.TIME -> timeRequest(message)
            Type.LAPS -> single(getLapMessages(robots).toMutableList())
            Type.LAP_MAN -> lapManual(message)
            Type.PIT_STOP -> pitStop(message)
            Type.FRAME -> frame(message)
            else -> throw LapsCounterException("Method not supported: [${message.type}]")
        }

    fun afterConnectionEstablished() = listOf(getState(), timeMessage())

    private fun command(message: Message): MessageResult {
        val parsedState = message.state ?: throw LapsCounterException("State is null")
        if (!isCorrectCommand(parsedState)) {
            throw LapsCounterException("Wrong current state to apply command: [$state]->[$parsedState]")
        }
        if (parsedState == state) {
            throw LapsCounterException("State already set up: [$state]->[$parsedState]")
        }

        val messageResult = MessageResult(ResponseType.BROADCAST)
        state = parsedState
        messageResult.add(getState())
        when (state) {
            State.READY -> {
                stopwatch.reset()
                robots.forEach { it.reset() }
                frameProcessor.reset()
                sortRobotsByLapsAndTime()
                messageResult.addAll(getLapMessages(robots))
            }

            State.STEADY -> {
            }

            State.RUNNING -> {
                stopwatch.start()
                if (raceTimeLimit > 0) {
                    lapsCounterScheduler.addSchedulerForFinishRace(TimeUnit.SECONDS.toMillis(raceTimeLimit))
                }
            }

            State.FINISH -> {
                stopwatch.finish()
                lapsCounterScheduler.resetSchedulers()
            }
        }

        messageResult.add(timeMessage())
        return messageResult
    }

    private fun isCorrectCommand(newState: State) =
        if (newState.ordinal == state.ordinal + 1) true
        else newState.ordinal == 0 && state.ordinal == State.entries.size - 1

    private fun robotInit(message: Message): MessageResult {
        val serial = message.serial ?: throw LapsCounterException("Robot serial is not defined")
        val existing = findRobot(serial)
        val robot: Robot =
            if (existing != null) {
                log.info("Reconnect robot {}", existing)
                existing
            } else {
                Robot(
                    serial = serial,
                    name = message.name ?: ("Robot $serial"),
                    num = maxNum() + 1,
                    place = robots.size + 1,
                ).also {
                    it.reset()
                    robots.add(it)
                    frameProcessor.robotInit(it.serial)
                    log.info("Connect new robot {}", it)
                }
            }
        log.debug("Connected robots: {}", robots)
        return broadcast(getLap(robot))
    }

    private fun maxNum() = robots.maxOfOrNull { it.num } ?: 0

    private fun robotEdit(message: Message): MessageResult {
        val robot = findRobotOrElseThrow(message.serial)
        log.info("Edit robot {}", robot)
        robot.name = message.name
        return broadcast(getLap(robot))
    }

    private fun robotRemove(message: Message): MessageResult {
        val robot = findRobotOrElseThrow(message.serial)
        robots.remove(robot)
        frameProcessor.robotRemove(robot.serial)
        return broadcast(listOf(message) + getLapMessages(sortRobotsByLapsAndTime()))
    }

    private fun timeRequest(message: Message): MessageResult =
        if (message.raceTimeLimit != null && state != State.RUNNING) {
            raceTimeLimit = message.raceTimeLimit
            broadcast(timeMessage())
        } else single(timeMessage())

    private fun lapManual(message: Message): MessageResult {
        if (state != State.RUNNING) {
            log.info("Lap manual ignored: state is not running")
            throw LapsCounterException("Lap manual ignored: state is not running")
        }
        if (message.laps == null) {
            throw LapsCounterException("Laps count is not defined")
        }

        val robot = findRobotOrElseThrow(message.serial)
        val affectedRobots = if (message.laps > 0) {
            incLaps(robot, stopwatch.time())
        } else {
            decLaps(robot)
        }
        return broadcast(getLapMessages(affectedRobots))
    }

    fun sortRobotsByLapsAndTime(): List<Robot> {
        val comparator = Comparator.comparingInt(Robot::laps).reversed()
            .thenComparingLong(Robot::time)
            .thenComparingInt(Robot::num)
        robots.sortWith(comparator)
        return robots.mapIndexedNotNull { index, robot ->
            robot.takeIf { robot.place != index + 1 }
                .also { robot.place = index + 1 }
        }
    }

    private fun pitStop(message: Message): MessageResult {
        if (state != State.RUNNING) {
            log.debug("PitStop ignored: state is not running")
            throw LapsCounterException("PitStop ignored: state is not running")
        }
        val robot = findRobotOrElseThrow(message.serial)

        robot.pitStopFinishTime = stopwatch.time() + pitStopTime
        log.info("Used PIT_STOP for robot: {}", robot.serial)

        val pitStopFinish = Message(Type.PIT_STOP_FINISH, serial = robot.serial)
        lapsCounterScheduler.addSchedulerForPitStop(pitStopFinish, pitStopTime)

        val pitStopStart = Message(Type.PIT_STOP, serial = robot.serial)
        return broadcast(listOf(pitStopStart, getLap(robot)))
    }

    private fun frame(message: Message): MessageResult {
        if (state != State.RUNNING) {
            log.debug("Frame ignored: state is not running")
            return MessageResult(ResponseType.SINGLE)
        }
        if (message.frame == null) {
            log.debug("Frame is null")
            return MessageResult(ResponseType.SINGLE)
        }
        val robot = findRobotOrElseThrow(message.serial)

        val raceTime = stopwatch.time()
        val frameType = frameProcessor.checkFrame(robot.serial, message.frame, raceTime)
        return when (frameType) {
            Type.LAP -> {
                val robots = incLaps(robot, raceTime)
                broadcast(getLapMessages(robots))
            }

            Type.LAP_MINUS -> {
                val robots = decLaps(robot)
                broadcast(getLapMessages(robots))
            }

            Type.FRAME -> {
                if (frameProcessor.isStartFrame(message.frame)) {
                    robot.currentLapStartTime = raceTime
                }
                single(Message(Type.FRAME))
            }

            else -> MessageResult(ResponseType.SINGLE)
        }
    }

    private fun incLaps(robot: Robot, raceTime: Long): List<Robot> {
        robot.incLaps(raceTime)
        return sortRobotsByLapsAndTime().ifEmpty {
            listOf(robot)
        }
    }

    private fun decLaps(robot: Robot): List<Robot> {
        robot.decLap()
        return sortRobotsByLapsAndTime().ifEmpty {
            listOf(robot)
        }
    }

    private fun getLapMessages(robots: List<Robot>) = robots.map { getLap(it) }

    fun scheduled(): Message =
        timeMessage().also {
            log.debug("Send time")
        }

    fun getState(): Message = Message(Type.STATE, state = state)

    fun timeMessage() = Message(Type.TIME, time = stopwatch.time(), raceTimeLimit = raceTimeLimit)

    private fun getLap(robot: Robot) =
        Message(
            type = Type.LAP,
            serial = robot.serial,
            name = robot.name,
            num = robot.num,
            laps = robot.laps,
            time = robot.time,
            bestLapTime = robot.bestLapTime,
            lastLapTime = robot.lastLapTime,
            place = robot.place,
        ).apply {
            robot.pitStopFinishTime?.let { pitStopFinishTime = it }
        }

    private fun findRobotOrElseThrow(serial: Int?) =
        findRobot(serial)
            ?: throw LapsCounterException("Cannot find robot by serial $serial")

    private fun findRobot(serial: Int?): Robot? =
        (serial ?: throw LapsCounterException("Robot serial is not defined"))
            .let { serialNotNull ->
                robots.find { it.serial == serialNotNull }
            }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(LapsCounterService::class.java)
    }
}
